### ISSUE TITLE
Ordering of signals matters

### ISSUE THREAD

**Issue by Cedrick84** [ReactiveKit/ReactiveKit#165]
Could have to do with  issue #153

Consider the following playground

```swift
import UIKit
import ReactiveKit
import Bond
import PlaygroundSupport

class Test {
    
    let startWithBool = SafePublishSubject<Bool>()
    let lastString = SafePublishSubject<String?>()
    
    init() {
        
        startWithBool
            .filter { $0 }
            .map { _ in return "test" }
            .bind(to: lastString)
        
        startWithBool
            .filter { !$0 }
            .eraseType()
            .with(latestFrom: lastString)
            .map { $0.1 }
            .bind(to: lastString)
        
        lastString
            .ignoreNil()
            .with(latestFrom: startWithBool)
            .observeNext(with: { (a, b) in
                print(b)
            })
    }
}


let test = Test()

test.lastString.observeNext { (_) in
    test.startWithBool.next(false)
}

test.startWithBool.next(true)

PlaygroundPage.current.needsIndefiniteExecution = true
```

The console logs nothing

When I change the order and put the `lastString.ignoreNil()..` block to the top I get the expected console output

```
true
false
```

I don't understand why it matters in what order I define these chains.
Help as usual is greatly appreciated.
Cheers

**Comment by srdanrasic** [comment]
A subject will propagate the received event to its observers in the order in which the observers were registered. This is ensured by the fix in #153. Prior to that the order was random. 

To answer you question, first check out `with(latestFrom:)` [example](http://rxmarbles.com/#withLatestFrom). 

Say that we have `a.with(latestFrom: b)`. A scenario that is relevant to your case is when `a` fires prior to `b`. As you can see in the diagram, no event will be fired in that case. It happens because  nothing "latest" exists yet.  There is no event from `b` to pair it with the event from `a`.

This behaviour is what causes the problem in your case. `lastString` fires an event, but `.with(latestFrom: startWithBool)` will generate nothing because `startWithBool` has not yet generated an event. Now, you might be wondering how's that because an event from `startWithBool` is what actually triggered the event on `lastString` (through first binding).

Let's inspect observers registered to `startWithBool`:

    0: observer from first `filter` operator
    1: observer from second `filter` operator
    2: observer from `with(latestFrom:)` operator

Given that, when you execute `test.startWithBool.next(true), i.e. send an event to `startWithBool` subject, the subject will propagate that event to its observers. That means that the event will first be propagated to the observer at index `0`. That will in turn trigger `filter`, `map`, `bind(to:)`, `ignoreNil`, `with(latestFrom:)` and `observeNext(with:)` operators! 

Notice that the whole chain including first binding and everything done to `lastString` is executed in one row. So, observer in `filter` receives the event, which is then immediately received by the observer in `map`, which is in turn received by binding. Binding sends the even to the `lastString` which in turn propagates the even to its own observers. 

Only when all that is done will the `startWithBool` subject propagate the initial event to the observer at index `1`! As you can see, `with(latestFrom:)` will already be executed before it receives the event from `startWithBool` - as that happens when observer at index `2` is given the event. 

Hope this helps with the understanding of the problem.

`with(latestFrom:)` operator is known to cause subtile issues and should be used with caution, especially when recursive / cyclic signals are involved. 

One way to solve the issue would be to dispatch part of the chain asynchronously so that all of the subject's observers are given the event. You can use the operator `observeOn` that basically dispatches events asynchronously on the given queue. Since the dispatch is asynchronous, you can dispatch it to any queue, including the one that the event is received on. Beware that the rest of that chain will be executed on given queue.

```swift
startWithBool
    .filter { $0 }
    .map { _ in return "test" }
    .observeOn(.main)
    .bind(to: lastString)
```

Now the binding and everything that follows (observations of `lastString`) will happen after `startWithBool` propagates the event to all observers. You would probably do this for the second binding as well.

You should also consider alternative approaches. Sometimes you can just flatMap signals instead of doing combinations with `with(latestFrom:)`.

**Comment by Cedrick84** [comment]
I owe you some beers! Cheers
